Clean up, simplify and generalize the Allen-key-probe code.

AnHardt opened the pull request:
Clean up, simplify and generalize the Allen-key-probe code.
- Add a optional feedrate parameter to `do_blocking_move()` and its companions.
  *Move Allen Key probe move scripts to their own functions
  There was an hidden extra move behind the scripts. This is configurable now.
  Added the extra moves to the relevant configs.
- Generalize and simplify triggered when stowed test
  The `PROBE_IS_TRIGGERED_WHEN_STOWED_TEST` is able to detect if a probe is
  successfully deployed/stowed.
  It was part of the Allen Key probes code only.
  Now the function is generally available.
  Currently this test is compulsory for Allen Key probes to not alter the used behavior
  If you want to add the test for an other type of probe simply add
  `#define PROBE_IS_TRIGGERED_WHEN_STOWED_TEST` to your config.
- Make use of the good features of a sled probe for all probes.
  Move and extend axis-unhomed test to deploy/stow_z_probe().
  Move and extend position store/restore to deploy/stow_z_probe().
  Now all kinds of probes can use the 'E' parameter in G29/M48.
  Allen key probes can be used now for grid and 3-point leveling.
  Deploying the Allen Key probe uses big moves in z direction.
  Too dangerous for an unhomed z-axis.
  Throw a compile time error when we try to configure an
  Allen Key probe homing to z-min and having no other z-min-endstop.
- Unify `deploy_z_probe()` and `stow_z_probe()` to `set_probe_deployed(bool deploy)`
  and give it a return value - `true` = failure - `false` = 'ok'.
  Let the callers return when a failure is detected.

AnHardt committed:
Add a optional feedrate parameter to `do_blocking_move()`

Add a optional feedrate parameter to `do_blocking_move()` and its companions.

AnHardt committed:
Move Allen Key probe move scripts to their own functions

There was an hidden extra move after the scripts. This is configurable now.
Added the extra moves to the relevant configs.

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -1780,215 +1780,244 @@ static void clean_up_after_endstop_or_probe_move() {
+    #if ENABLED(Z_MIN_PROBE_ENDSTOP)
+      #define _TRIGERED_WHEN_STOWED_TEST (READ(Z_MIN_PROBE_PIN) != Z_MIN_PROBE_ENDSTOP_INVERTING)
+    #else
+      #define _TRIGERED_WHEN_STOWED_TEST (READ(Z_MIN_PIN) != Z_MIN_ENDSTOP_INVERTING)
```
Might as well add the extra G here: `_TRIGGERED_WHEN_STOWED_TEST`

thinkyhead commented on Marlin/Marlin_main.cpp:
```diff
@@ -1780,215 +1780,244 @@ static void clean_up_after_endstop_or_probe_move() {
+  #define STOW_PROBE() deploy_stow_probe( true )
+
+  // returns false for ok and true for failure
+  static bool deploy_stow_probe(bool stow) {
```
I think we should switch this around and make the flag `deploy` instead of `stow` – so this follows the same pattern as `endstops.enabled` — that is, `true` when things are deployed or enabled. Then the function should be named differently as well…

``` cpp
static bool set_probe_deployed(bool deploy) { . . . }
```

thinkyhead commented on Marlin/SanityCheck.h:
```diff
@@ -425,16 +425,12 @@
+#if ENABLED(Z_PROBE_ALLEN_KEY) && (Z_HOME_DIR < 0) && ENABLED(Z_MIN_PROBE_USES_Z_MIN_ENDSTOP_PIN)
+  #error "You can't home to a z min endstop with a Z_PROBE_ALLEN_KEY"
 #endif
-
```
This empty line should be put back.

thinkyhead commented on Marlin/example_configurations/delta/biv2.5/Configuration.h:
```diff
@@ -538,6 +543,11 @@ const bool Z_MIN_PROBE_ENDSTOP_INVERTING = true; // set to true to invert the lo
+  #define Z_PROBE_ALLEN_KEY_STOW_4_X 0.0
+  #define Z_PROBE_ALLEN_KEY_STOW_4_Y 0.0
+  #define Z_PROBE_ALLEN_KEY_STOW_4_Z Z_PROBE_ALLEN_KEY_STOW_3_Z
+  #define Z_PROBE_ALLEN_KEY_STOW_4_FEEDRATE XY_PROBE_SPEED
```
So I see the last move is just a move back to center… I wonder if that should just be the standard.

AnHardt replied:
```
 -      destination[Z_AXIS] = Z_PROBE_ALLEN_KEY_STOW_3_Z;       
 -      prepare_move_to_destination();      
 -      
 -      // Home XY for safety       
 -      feedrate = homing_feedrate[X_AXIS] / 2;     
 -      destination[X_AXIS] = 0;        
 -      destination[Y_AXIS] = 0;        
 -      prepare_move_to_destination(); // this will also set_current_to_destination     
 -      
 -      feedrate = old_feedrate;        
```

Just a replacement for the formerly hard coded move. Now you can change or omit it - if you want.

thinkyhead replied:
Ja, I like the new configurable method better.

AnHardt committed:
Make use of the good features of a sled probe for all probes.

Move and extend axis_unhomed test to deploy/stow_z_probe().
Move and extend position store/restore to deploy/stow_z_probe().

Now all kinds of probes can use the 'E' parameter in G29/M48.
Allen key probes can be used now for grid and 3-point levelling.

Deploying the Allen Key probe uses big moves in z direction.
Too dangerous for an unhomed z-axis.
Throw a compile time error when we try to configure an
Allen Key probe homing to z-min and having no other z-min-endstop.

AnHardt committed:
Generalize and simplify 'triggered when stowed test`

The `PROBE_IS_TRIGGERED_WHEN_STOWED_TEST` is able to detect if a probe is
successfully deployed/stowed.
It was part of the Allen Key probes code only.
Now the function is generally available.

Currently this test is compulsory for Allen Key probes to not alter the used behaviour
If you want to add the test for an other type of probe simply add
```
#define PROBE_IS_TRIGGERED_WHEN_STOWED_TEST
```
to your config.

AnHardt commented:
Updated & rebased.
